registry/storage/garbagecollect.go

```diff
@@ -89,25 +100,31 @@ func MarkAndSweep(ctx context.Context, storageDriver driver.StorageDriver, regis
 				return fmt.Errorf("failed to retrieve manifest for digest %v: %v", dgst, err)
 			}
 
-			descriptors := manifest.References()
-			for _, descriptor := range descriptors {
+			for _, descriptor := range manifest.BlobReferences() {
 				markSet[descriptor.Digest] = struct{}{}
 				emit("%s: marking blob %s", repoName, descriptor.Digest)
 			}
 
-			return nil
-		})
+			for _, descriptor := range manifest.ManifestReferences() {
+				if _, ok := unmarkedManifests[descriptor.Digest]; ok {
+					delete(unmarkedManifests, descriptor.Digest)
+					manifestList = append(manifestList, descriptor.Digest)
+				}
+			}
+		}
 
-		// In certain situations such as unfinished uploads, deleting all
-		// tags in S3 or removing the _manifests folder manually, this
-		// error may be of type PathNotFound.
-		//
-		// In these cases we can continue marking other manifests safely.
-		if _, ok := err.(driver.PathNotFoundError); ok {
-			return nil
+		if len(unmarkedManifests) != 0 {
```

Comment:
If there are manifests to remove add them to the manifestArr and give a list of all tags that we need to make sure are no longer referenced in the history of any tag in the repository.